Fix for Issue 11285: warning message displayed when updating an existing image

shishir-a412ed opened the pull request:
Fixes #11285

shishir-a412ed commented:
Fixes #11285

tiborvass commented:
Putting in design review to make sure the solution to the issue is the correct one.

shishir-a412ed commented:
@jlhawn Let me know what do you think of replacing the existing image name to <imagename_id>:<tag> rather than <none>:<none>.

So for example if someone is loading busybox:latest with ID (4349ec6bf114) and there is already an existing image busybox:latest with ID (f7189223014c). 

The existing one should become busybox_f7189223014c:latest
and the new one busybox:latest

Shishir

jlhawn commented:
After looking at the patch, it looks like you just print the message to stdout on the daemon. We should probably use the regular logging package *and* figure out how to display it to the client (would require a refactor because the `(*TagStore).Set()` method doesn't take a job output/stream formatter argument.)

@shishir-a412ed as for your suggestion:

> Let me know what do you think of replacing the existing image name to : rather than :.
>
> So for example if someone is loading busybox:latest with ID (4349ec6bf114) and there is already an existing image busybox:latest with ID (f7189223014c).
>
> The existing one should become busybox_f7189223014c:latest
and the new one busybox:latest

hmm, changing the full name of the repository is kind of weird. Can we just change the tag to be the image ID? ex: `busybox:latest -> busybox:f7189223014c`. In fact, I think you can *already* refer to the image this way anyway. Let me try it... yep:

```bash
$ docker images busybox
REPOSITORY          TAG                 IMAGE ID            CREATED             VIRTUAL SIZE
busybox             latest              4986bf8c1536        10 weeks ago        2.433 MB
...
$ docker run --rm busybox:4986bf8c1536 echo 'it works!'
it works!
```

So, no need to do anything but display the info message to the user, in my opinion.

shishir-a412ed commented:
@jlhawn the reason for renaming is to avoid getting it named to <none>:<none>.
How do I know that <none>:<none> stands for a busybox image (older version) ?

Currently if I do a `docker images --all` I see a lot of <none>:<none> images which resulted due to name conflicts. I think `busybox:latest -> busybox: f7189223014c` should work, although I have to implement that part.

Do you want that change to be a part of separate PR or just add another commit in this PR ?
I ll take a look at the logging facility to update the warning message.

shishir-a412ed commented:
@jlhawn I discussed with my QE team, warning message is good enough for our problem.
I have redirected the warning message to the client side.

Please take a look and let me know if this works for you.

Shishir

jlhawn commented on graph/tags.go:
```diff
@@ -222,6 +223,15 @@ func (store *TagStore) Set(repoName, tag, imageName string, force bool) error {
+		}
+
 		repo = r
 		if old, exists := store.Repositories[repoName][tag]; exists && !force {
```
Looks like there's already a check here, you should be able to re-use this block of code.

also note that it's not necessary here to use `store.Repositories[repoName]` as this variable was already declared/set in the outer `if` statement as `r`.

and let's not special case the message just for the "load" job. It's useful to have in general. We can also get rid of the `job` argument to this method and just use an `io.Writer`.

shishir-a412ed replied:
If you are talking about line 227, this one is the existing upstream code.
If you like I can change this to r[tag] instead of store.Repositories[repoName][tag] and see if that works.

shishir-a412ed commented:
@jlhawn I have modified the code.

The reason I was receiving `job *engine.Job` as an argument, and not `job.Stdout`, and putting a check just on "load" was because it was throwing a lot of errors in the test suite. Probably it was running this warning message for other use cases and breaking the expected behaviour.

I have modified the code to receive io.Writer. Lets wait and see if Jenky breaks.

Shishir

jlhawn commented:
Design looks good to me: the "tag set" job can now have messages written to its stdout and this is now used to display a message to the Client when there's a conflict in `docker load`.

Code changes LGTM as well.

@tiborvass please take a look and move into code-review if the design is okay with you.

shishir-a412ed commented:
@LK4D4 I have made the requested changes, what we discussed on irc.

Adding @jlhawn too. I have seen this jenkins error before on this PR. This doesn't seem to be related to this PR. Let me know if it is and I ll take a look.

jlhawn commented:
@shishir-a412ed I triggered a rebuild and janky passed ;-)

shishir-a412ed commented:
@jlhawn Awesome !!

LK4D4 commented:
@shishir-a412ed Super nice that you reduced your diff.

LK4D4 commented:
ping @tiborvass

shishir-a412ed commented:
@tiborvass Can we merge this ?

icecrime commented:
Moving to code review (hoping @tiborvass won't have any objection).

icecrime commented on graph/tags.go:
```diff
@@ -241,8 +246,15 @@ func (store *TagStore) Set(repoName, tag, imageName string, force bool) error {
+
+			if old != img.ID && out != nil {
+				str := "The image " + repoName + ":" + tag + " already exists, renaming the old one with id " + old[:12] + " to <none>:<none>\n"
+				out.Write([]byte(str))
```
Am I missing something or `out` can be `nil` here (especially when called from `TagStore.Set`)?

tiborvass replied:
this should be `fmt.Fprintf(out, "The image %s:%s already exists, renaming the old one with ID %s to empty string\n", repoName, tag, old[:12])`

shishir-a412ed replied:
@tiborvass I have made the change. Please check now.

icecrime commented on graph/load.go:
```diff
@@ -68,7 +68,7 @@ func (s *TagStore) CmdLoad(job *engine.Job) engine.Status {
 		for imageName, tagMap := range repositories {
 			for tag, address := range tagMap {
-				if err := s.Set(imageName, tag, address, true); err != nil {
+				if err := s.SetLoad(imageName, tag, address, true, job.Stdout); err != nil {
```
Sorry but I don't get it: if `SetLoad` returns an error, why isn't it reported to the client? Why do we need to pass the `job.Stdout` and emit a new line of output there, isn't the solution just to write the error result to `w` in `postImagesLoad`?

shishir-a412ed replied:
@icecrime it's a warning message not an error. We still want the `docker load` to go through.

icecrime replied:
But does it mean that when it doesn't go through (such as needing `force`), we're gonna get both a warning and an error?

shishir-a412ed replied:
@icecrime it depends ... if the tar ball being loaded has a new content i.e image ID's are different, but the image name: tag of the old and new images are same.

The idea is to show a display message to user, so that he doesn't get confused when the old image with the same name is renamed to none:none.

Commit:
Applying: Fix for Issue 11285: warning message displayed when updating an existing image

Signed-off-by: Shishir Mahajan <shishir.mahajan@redhat.com>

tiborvass commented on graph/tags.go:
```diff
@@ -219,6 +220,10 @@ func (store *TagStore) Delete(repoName, ref string) (bool, error) {
+	return store.SetLoad(repoName, tag, imageName, force, nil)
+}
+
+func (store *TagStore) SetLoad(repoName, tag, imageName string, force bool, out io.Writer) error {
```
@shishir-a412ed why not make these one `Set` method ? Just add the `out` param.

shishir-a412ed replied:
@LK4D4 would you like to answer @tiborvass ?

tiborvass replied:
lol

dmcgowan commented on graph/tags.go:
```diff
@@ -241,8 +246,16 @@ func (store *TagStore) Set(repoName, tag, imageName string, force bool) error {
+				fmt.Fprintf(out, "The image %s:%s already exists, renaming the old one with ID %s to empty string\n", repoName, tag, old[:12])
+
+			}
+			if !force {
```
Why is this check after the output statement since no changes are saved if this condition is true?

shishir-a412ed replied:
@dmcgowan what do you mean ? Can you elaborate a bit more.
Do you mean why display the warning message and then throw the error too if `force` flag is false ?

dmcgowan replied:
> Do you mean why display the warning message and then throw the error too if force flag is false ?

Yes, when the error is thrown, it doesn't appear anything is renamed

shishir-a412ed replied:
@dmcgowan I have made the changes.

dmcgowan commented:
Tested and I am seeing the error message, I am confused on how the force flag works since `docker load` does not have a force flag.

~~~
Usage: docker load [OPTIONS]

Load an image from a tar archive on STDIN

  --help=false       Print usage
  -i, --input=       Read from a tar archive file, instead of STDIN
~~~

When I just run, it seems to work.  Not sure how to test "unforced" though

~~~
$ docker load < /tmp/foobar.tar
The image foobar:latest already exists, renaming the old one with ID 883f763834c1 to empty string
~~~

shishir-a412ed commented:
@dmcgowan AFAIK docker tag command also makes a call to the `Set` method, and it has the `force` flag.

dmcgowan commented:
Alright, thanks for the clarification.  Is it expected that issuing a `-f` on tag won't show the message you added?

~~~
$ docker tag 883f763834c1 foobar
FATA[0000] Error response from daemon: Conflict: Tag latest is already set to image 6705fbab54cee055f6ef0239709284a024f8dd8b30c102d6596dcd405779fa2e, if you want to replace it, please use -f option
$ docker tag -f 883f763834c1 foobar
$
~~~

shishir-a412ed commented:
@dmcgowan yes this change is only related to `docker load`

dmcgowan commented:
LGTM

dmcgowan commented:
ping @tiborvass

tiborvass commented:
LGTM
